lib/migration.js

```diff
@@ -412,92 +397,162 @@ function mixinMigration(MySQL, mysql) {
-      var newFksStatement = self.getAlterStatement(model, addFksSql);
-      var stmtList = [alterStmt, newFksStatement].filter(function(s) { return s.length; });
-
-      //set up an object to pass back all changes, changes that have been run,
-      //and foreign key statements that haven't been run
-      var retValues = {
-        statements: stmtList,
-        query: stmtList.join(';'),
-        newFks: newFksStatement,
-      };
-
-      //if we're running in read only mode OR if the only changes are foreign keys additions,
-      //then just return the object directly
-      if (checkOnly || !alterStmt.length) {
-        done(null, true, retValues);
-      } else {
-        //if there are changes in the alter statement, then execute them and return the object
-        self.execute(alterStmt, function(err) {
-          done(err, true, retValues);
-        });
-      }
-    } else {
-      done();
+  MySQL.prototype.applySqlChanges = function(model, pendingChanges, cb) {
+    var self = this;
+    var thisQuery = 'ALTER TABLE ' + self.tableEscaped(model) + ' ' + pendingChanges.toString();
+    thisQuery = thisQuery + ';';
+    if (pendingChanges.length) {
+      self.execute(thisQuery, cb);
     }
+  };
 
-    function actualize(propName, oldSettings) {
-      var newSettings = m.properties[propName];
-      if (newSettings && changed(newSettings, oldSettings)) {
-        var pName = self.client.escapeId(propName);
-        sql.push('CHANGE COLUMN ' + pName + ' ' + pName + ' ' +
-          self.buildColumnDefinition(model, propName));
+  MySQL.prototype.getForeignKeySQL = function(model, actualFks) {
+    var self = this;
+    var m = this.getModelDefinition(model);
+    var addFksSql = [];
+
+    if (actualFks) {
+      var keys = Object.keys(actualFks);
+      for (var i = 0; i < keys.length; i++) {
+        var constraint = self.buildForeignKeyDefinition(model, keys[i]);
+
+        if (constraint) {
+          addFksSql.push('ADD ' + constraint);
+        }
       }
     }
+    return addFksSql;
+  };
 
-    function changed(newSettings, oldSettings) {
-      if (oldSettings.Null === 'YES') {
-        // Used to allow null and does not now.
-        if (!self.isNullable(newSettings)) {
-          return true;
+  MySQL.prototype.addForeignKeys = function(model, fkSQL, cb) {
+    var self = this;
+    var m = self._models[model];
+
+    if ((!cb) && ('function' === typeof fkSQL)) {
+      cb = fkSQL;
+      fkSQL = undefined;
+    }
+
+    if (!fkSQL) {
+      var newFks = m.settings.foreignKeys;
+      if (newFks)
+        fkSQL = self.getForeignKeySQL(model, newFks);
+    }
+    if (fkSQL && fkSQL.length) {
+      self.applySqlChanges(model, fkSQL, function(err, result) {
+        if (err) cb(err);
+        else
+          cb(null, result);
+      });
+    } else cb(null, {});
+  };
+
+  MySQL.prototype.dropForeignKeys = function(model, actualFks, cb) {
+    var self = this;
+    var m = self._models[model];
```

Comment:
I think use the original `var m = this.getModelDefinition(model)` would be better, `getModelDefinition` is defined here:
https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/model-builder.js#L87
From the implementation they may return a different thing